File: c/collatz-conjecture/collatz_conjecture.c
#include "collatz_conjecture.h"

int steps(int start) {
    int result_steps = 0;

    if (start <= 0) {
        return ERROR_VALUE;
    }

    while(start != 1) {
        if (start % 2 == 0) {
            start /= 2;
        } else {
            start = 3 * start + 1;
        }
        result_steps++;
    }
    return result_steps;
}

File: go/collatz-conjecture/collatz_conjecture.go
package collatzconjecture

import "errors"

func CollatzConjecture(n int) (int, error) {
	if n <= 0 {
		return 0, errors.New("n must be at least 1")
	}
	return collatz(n, 0), nil
}

func collatz(n, steps int) int {
	switch {
	case n == 1:
		return steps
	case n%2 == 0:
		return collatz(n/2, steps+1)
	default:
		return collatz(3*n+1, steps+1)
	}
}
